File: drills/ts/solutions/nc100_edit_distance__memoised.ts
export function minDistance(word1: string, word2: string): number {
  const memo = new Map<string, number>();

  // The same three edits as an explicit choice from the front. Running out of
  // one word costs whatever is left of the other, since every remaining
  // character has to be inserted or deleted.
  const cost = (i: number, j: number): number => {
    if (i >= word1.length) return word2.length - j;
    if (j >= word2.length) return word1.length - i;
    const key = `${i},${j}`;
    if (!memo.has(key)) {
      memo.set(
        key,
        word1[i] === word2[j]
          ? cost(i + 1, j + 1)
          : 1 + Math.min(cost(i + 1, j + 1), cost(i + 1, j), cost(i, j + 1)),
      );
    }
    return memo.get(key)!;
  };

  return cost(0, 0);
}

File: drills/go/solutions/nc100_edit_distance__memoised.go
package main

func minDistance(word1 string, word2 string) int {
	memo := map[[2]int]int{}
	var from func(i, j int) int
	from = func(i, j int) int {
		if i == len(word1) {
			return len(word2) - j
		}
		if j == len(word2) {
			return len(word1) - i
		}
		key := [2]int{i, j}
		if v, ok := memo[key]; ok {
			return v
		}
		if word1[i] == word2[j] {
			memo[key] = from(i+1, j+1)
		} else {
			memo[key] = 1 + min(from(i+1, j+1), min(from(i+1, j), from(i, j+1)))
		}
		return memo[key]
	}
	return from(0, 0)
}
